FlowCryptAppTests/Functionality/Services/Key Services/KeyServiceTests.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  KeyServiceTests.swift
+//  FlowCryptAppTests
+//
+//  Created by  Ivan Ushakov on 15.10.2021
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+    
+
+import XCTest
+@testable import FlowCrypt
+
+final class KeyServiceTests: XCTestCase {
+
+    func testGetSigningKeyFirstEmail() throws {
+        // arrange
+        let keyStorage = KeyStorageMock()
+        let keyInfo = KeyInfo()
+        keyInfo.user = UserObject(name: "Bill", email: "bill@test.com", imap: nil, smtp: nil)
+        keyStorage.keysInfoResult = {
+            [keyInfo]
+        }
+
+        let passPhraseService = PassPhraseServiceMock()
+        passPhraseService.passPhrases = [
+            PassPhrase(value: "phrase", fingerprints: ["2"], date: nil)
+        ]
+
+        let userEmail = "bill@test.com"
+
+        let keyDetails_1 = KeyDetails(
+            public: "public1",
+            private: "private1",
+            isFullyDecrypted: nil,
+            isFullyEncrypted: nil,
+            ids: [KeyId(longid: "1", fingerprint: "1")],
+            created: 0,
+            lastModified: nil,
+            expiration: nil,
+            users: ["Joe <joe@example.com>", "Bill <bill@test.com>"],
+            algo: nil
+        )
+        let keyDetails_2 = KeyDetails(
+            public: "public2",
+            private: "private2",
+            isFullyDecrypted: nil,
+            isFullyEncrypted: nil,
+            ids: [KeyId(longid: "2", fingerprint: "2")],
+            created: 0,
+            lastModified: nil,
+            expiration: nil,
+            users: ["Bill <bill@test.com>", "Jim <jim@example.com>"],
+            algo: nil
+        )
+        let keyParser: KeyService.KeyParser = { data in
+            return CoreRes.ParseKeys(
+                format: .armored,
+                keyDetails: [keyDetails_1, keyDetails_2]
+            )
+        }
```

Comment:
There is no need to create a KeyParser type and mock this: this mock can be avoided by instead generating the actual keys with desired properties. That will be a much more direct test of the functionality, and it will simplify production code by removing `KeyService.KeyParser` which is there only for these tests.

Reply:
Idea was to have pure unit test not integration one. Since we have separate test for the parsing inside Core class there is no need to test it again. Also this makes test slower because parsing and executing TS code takes time. But I can use your approach if you want.

Reply:
Thank you - I'm not too worried between a 5ms test and 30ms test. What worries me is this: 

https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.giphy.com%2Fmedia%2FsYz9N4FHOLlxC%2Fgiphy.gif&f=1&nofb=1

And also complex production code.

For our codebase, mocks are more appropriate for things that are really heavy like network calls, or situations where it's not possible to easily induce the desired test behavior without a mock.

But for a bit of parsing, it's better to let the production code do its job, even if it's already covered by other tests.

Reply:
I tried to use `Core.generateKey` but private key always have only one KeyDetails for there is no way to check algorithm completely like I did in my original test. Maybe it is not possible to have several KeyDetails for the one armored private key?

Reply:
One armored private key gives you a single KeyDetails, but nothing stops you from making several private keys. After all, in the app storage, there is one private key per database record, and they all get parsed separately, each producing one KeyDetails structure.